lti_consumer/plugin/views.py

```diff
@@ -501,3 +512,398 @@ def list(self, *args, **kwargs):
                 "error": "above_response_limit",
                 "explanation": "The number of retrieved users is bigger than the maximum allowed in the configuration.",
             }, status=HTTP_403_FORBIDDEN)
+
```

Comment:
the import renaming in exams is a bit of an indicator of weird stuff:
from lti_consumer.plugin.views import start_proctoring as lti_start_proctoring

would we expect any app to ever be able to use these views without wrapping? If not can we get any benefit by having them not be structured as views? They seem to all do request param copying and the exam wrapper also does that, so we make a copy of a copy to add things in layers. Maybe these functions could just take a dictionary of extra stuff to add. Maybe that could go all the way down into proctoring_preflight because all of its callers do the same setup

Reply:
Do we absolutely need edx-exams to be providing the `attempt_number` and such? Could the UI just provide these values when starting the flow? In that way we can leave the views entirely within the lti-consumer library and avoid wrapping or using api functions right?

Reply:
What would that look like? The `launch_gate_endpoint_proctoring` view should be invoked when the Tool makes a third party authentication request to the Platform (i.e. `edx-exams`). How would the frontend pass these data into the flow? I could see how the frontend could pass `attempt_number` when it kicks off the LTI flow (via the `start_proctoring` view) when, say, a learner presses a "start" button, but that data isn't needed the LTI flow is being kicked off. Once the flow is kicked off, will the frontend even be involved until it comes time to render content?

Reply:
I'm going to call "LTI views" the endpoints from `xblock-lti-consumer` that are installed into `edx-exams`.

For the first request that is made, which is the third party login initiation (i.e. `start_proctoring`), the frontend _could_ make a request to the LTI view that would include the LTI `config_id`. That makes sense to me. I assume the frontend would get it in some backend-for-frontend request before rendering the exam, right?

The Platform receives this request, grabs the `LtiConfiguration` using the `config_id`, crafts a preflight URL using the `LtiProctoringConsumer`, and returns a redirect to that preflight URL. The browser makes the request. Next, the Tool gets the request and makes a third party authentication request to the LTI view through the browser. `edx-exams` gets the request and has to issue an LTI launch. At this point, the LTI view needs to know `attempt_number`, `resource_link`, etc. in order to issue a launch. These data come from `edx-exams`, but the LTI view has no knowledge of them. I don't think the frontend can help here, unless I'm mistaken?

Reply:
@ashultz0 

> would we expect any app to ever be able to use these views without wrapping? If not can we get any benefit by having them not be structured as views? They seem to all do request param copying and the exam wrapper also does that, so we make a copy of a copy to add things in layers. Maybe these functions could just take a dictionary of extra stuff to add. Maybe that could go all the way down into proctoring_preflight because all of its callers do the same setup

Unless we can find an alternative way to pass the necessary data around, I don't think so. I can see what that your suggestion would look like. You're envisioning that the input to any of these function would be a dictionary (not a request), and then the output would be a response object, right?

Reply:
I think I meant instead of each layer passing in a request and a dict and combining the dict into the request params to make new request params we would just pass the request and ever increasing set of new params down to the lowest level before sticking them together and getting the response. The various layers all have the view interface but... is anyone ever really going to use the inner views as views? If not, they don't have the view argument set.

Reply:
> I'm going to call "LTI views" the endpoints from `xblock-lti-consumer` that are installed into `edx-exams`.
> 
> For the first request that is made, which is the third party login initiation (i.e. `start_proctoring`), the frontend _could_ make a request to the LTI view that would include the LTI `config_id`. That makes sense to me. I assume the frontend would get it in some backend-for-frontend request before rendering the exam, right?
> 
> The Platform receives this request, grabs the `LtiConfiguration` using the `config_id`, crafts a preflight URL using the `LtiProctoringConsumer`, and returns a redirect to that preflight URL. The browser makes the request. Next, the Tool gets the request and makes a third party authentication request to the LTI view through the browser. `edx-exams` gets the request and has to issue an LTI launch. At this point, the LTI view needs to know `attempt_number`, `resource_link`, etc. in order to issue a launch. These data come from `edx-exams`, but the LTI view has no knowledge of them. I don't think the frontend can help here, unless I'm mistaken?

I guess I was under the impression that the contextual details such as attempt and resource_id wood be somehow be passed through the initial authentication steps. Otherwise once we do get to the step where the tool requests a launch from edx-exams how would edx exams even know what specific attempt this is for if it's only context at that point is a certain LtiConfiguration?

Reply:
They're not, unfortunately.

The specification doesn't detail how to accomplish this. It's a good question. I agree that it's not really clear how to get much more than the currently requesting user and the provider being used for the exam (from the `client_id`) when crafting the LTI launch message.

I believe that the way the existing LTI 1.3 launch gets around this is by being coupled to the XBlock `location` and by misusing the `login_hint` parameter. The `login_hint` parameter refers to a `location` on the `LtiConfiguration`, but it also lets the library pull data about the context of the LTI launch through the XBlock. I think that's why the full decoupling of the LTI views from `LtiConfiguration.location` has been challenging. It's a little bit like apples and oranges, but the `context_claim` is similar to `attempt_number` or `resource_link`. I considered whether we could abuse the `login_hint` to refer to an attempt to tie the two requests together, but it's really not the purpose of the parameter. Nevertheless, we _could_ do that, but it doesn't address the other data that the library will need (e.g. `resource_link`, user `role`, etc.).

I'll think through some options tomorrow and report back.